contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,220 @@
+
+## What Is a Module?
+
+You might often encounter the term such as *module, package, library, dependency, plugin,* etc. when developing your own application in any programming languages, including JavaScript. We will not dive into the exact definition of each term (because that is not the scope of this article), but we will define a module as a collection of code that share the common purpose or support a certain functionality that is placed in different file.
+
+Let's say you have an application with functionalities A and B, where functionality A needs functionality B. Without modules, both of these functionalities are mixed together in the code base without a clear separation. With modules, we can separate those 2 functionalities into separate module each. When A need B, A will "include" B and A will be able to work as if A and B has never been separated. The exact usage of modules will be illustrated in the **How to Modularize JavaScript Code** section after the following section.
+
+## Why Use Modules in JavaScript?
+
+Following are reasons why use modules in JavaScript. The reasons below are not exclusive to JavaScript, but may also be general reasons for using modules in other languages.
+
+**1. Better management of code base** <br>
+Using modules will break the code base into smaller parts which if done well can help in managing a code base, especially a large one. To be able to modularize your code well, you can apply Software Engineering principles such as [SOLID](http://se-education.org/se-book/principles/solidPrinciples/index.html), etc.
+
+**2. Namespacing** <br>
+Because of how [JavaScript scoping](https://www.sitepoint.com/demystifying-javascript-variable-scope-hoisting/) works, avoiding global variable can be an issue. Modules allows JavaScript developer to avoid this problem by isolating the namespace to a specific module.
+
+**3. Reusability** <br>
+As per [Don't Repeat Yourself (DRY) Principle](http://se-education.org/se-book/principles/dryPrinciple/index.html), modules allows developers to reuse their code that is contained in a module. If for example we have an application that rely on a fuctionality such as string comparison function, we can separate that function into a module and let the application to use the function from that module instead of having to always repeating that function in all the places where it is needed.
+
+## How to Modularize JavaScript Code?
+
+There are 3 most common implementations of modules in JavaScript: *ES6 modules* as the most recent and official implementation, *CommonJS* which is used by NodeJS, and the old *module pattern*. This article will cover all those 3 implementations because there is still a large number of applications using those 3 implementations.
```

Comment:
Suggestion:

> There are 3 common ways to use modules in JavaScript: 1. using _ES6_ modules, 2. using _CommonJS_, 3.  using the _module pattern_.  While ES6 is the most recent and the 'official' approach, this article covers the other two as well because there is still a large number of existing projects that use them.